src/dune/build_system.ml

```diff
@@ -328,10 +328,11 @@ type t =
       Fdecl.t
   ; (* Package files are part of *)
     packages : (Path.Build.t -> Package.Name.Set.t) Fdecl.t
-  ; caching : caching option
+  ; mutable caching : caching option
   ; sandboxing_preference : Sandbox_mode.t list
   ; mutable rule_done : int
   ; mutable rule_total : int
+  ; vcs : Vcs.t list Fdecl.t
```

Comment:
Do I understand correctly that the field `caching` can be changed many times, while `vcs` is set only once? It's not immediately clear why these two fields use different mutation mechanisms.

Reply:
The cache module uses an immutable model, were the new cache is returned by every operation. `Build_system.t` does not, but that's pretty much out of my scope here :)

Reply:
I'm not sure what you mean by "by every operation". In any case, I'd appreciate if you add a comment to each of the new fields saying when they are mutated.

Reply:
(I'll most likely need to rewrite this code to make it mutation free for the file-watching mode of Dune, so such comments would be very useful.)

Reply:
What's itching me a bit TBH is that this is not related to the PR : this one simply adds a neat immutable list of VCS to the build system instance :) There certainly are other mutable fields, some of them years old, because the choice was made to have a global Build_system.t instance which entails that we sometime need to mutate some fields. I don't mind adding a comment suggesting we should redesign this - I definitely think we do - but that seems a bit out of scope here.

Reply:
Mutability is difficult to reason about, so ideally all such fields should be documented. In most cases I need to do detective work to figure out what's the right way to get rid of mutable state but in this case you might have the right ideas in your head already, so please share them -- that would save me some time in future! (And if I come back to you with questions about this piece of code in a month time, you'll most likely have forgotten about it already.)